src/linker/Linker.Dataflow/ValueNode.cs

```diff
@@ -1239,23 +1252,36 @@ public override bool Equals (ValueNode other)
 				return false;
 
 			ArrayValue otherArr = (ArrayValue) other;
-			return Size.Equals (otherArr.Size);
+			bool equals = Size.Equals (otherArr.Size);
+			equals &= IndexValues.Count == otherArr.IndexValues.Count;
+			if (!equals)
+				return false;
+
+			// If both sets T and O are the same size and "T intersect O" is empty, then T == O.
+			HashSet<KeyValuePair<int, ValueBasicBlockPair>> thisValueSet = new (IndexValues);
+			HashSet<KeyValuePair<int, ValueBasicBlockPair>> otherValueSet = new (otherArr.IndexValues);
+			thisValueSet.ExceptWith (otherValueSet);
+			return thisValueSet.Count == 0;
 		}
 
 		protected override string NodeToString ()
 		{
-			return ValueNodeDump.ValueNodeToString (this, Size);
+			// TODO: Use StringBuilder and remove Linq usage.
+			return $"(Array Size:{ValueNodeDump.ValueNodeToString (this, Size)}, Values:({string.Join (',', IndexValues.Select (v => $"({v.Key},{ValueNodeDump.ValueNodeToString (v.Value.Value)})"))})";
 		}
 
 		protected override IEnumerable<ValueNode> EvaluateUniqueValues ()
 		{
 			foreach (var sizeConst in Size.UniqueValuesInternal)
-				yield return new ArrayValue (sizeConst);
+				yield return new ArrayValue (sizeConst, ElementType, IndexValues);
 		}
 
 		protected override ValueNode ChildAt (int index)
 		{
 			if (index == 0) return Size;
+			if (index - 1 <= IndexValues.Count)
+				return IndexValues[index].Value;
+
 			throw new InvalidOperationException ();
```

Comment:
Should we hard-fail in this case? I would expect us to return unknown value here.